openhtf/plugs/frontline/bluetooth_sniffer.py

```diff
@@ -0,0 +1,277 @@
+  """Raised when a BLE sniffer command fails."""
+  def __init__(self, command, response):
+    super(SnifferError, self).__init__(
+        'Command "{}" failed with response "{}"'.format(command, response))
+    self.response = response
+
+
+class Sniffer(object):
+  """Class for controlling a Frontline bluetooth sniffer via Telnet.
+
+  Args:
+    host: Hostname or IP address of the sniffer's automation server.
+    port: TCP port fort the sniffer's automation server.
+    log_path: Path on the sniffer's automation server at which to save logs.
+  """
+  def __init__(self,
+               host='localhost',
+               port=DEFAULT_AUTOMATION_PORT,
+               log_path=REMOTE_LOG_PATH):
+    self._address = host, port
+    self._log_path = log_path
+
+    self.capture_id = None
+    self.capture_start_time = None
+    self._files = []
+    self._socket = None
+
+  def _connect(self):
+    """Open a connection to the sniffer's automation server."""
+    _LOG.info('Connecting to sniffer server at %s:%d', *self._address)
+    assert self._socket is None
+    self._socket = socket.create_connection(self._address, 10)
+
+    _LOG.info('Stopping any previously running sniffer captures')
+    self._socket.settimeout(16)
+    try:
+      self._sniffer_command('Stop FTS')
+    except (SnifferError, socket.timeout):
+      pass
+
+    self._socket.settimeout(None)
+
+  def _sniffer_command(self, command, *args):
+    """Send command to the sniffer's automation server and wait for response."""
+    full_command = ';'.join((command,) + args)
+    _LOG.debug('Sniffer command: %s', full_command)
+    self._socket.sendall(full_command.encode() + b'\n')
+
+    for line in self._socket.makefile():
+      response = line.split(';')
+      _LOG.debug('Sniffer response: %s', line.strip())
+      if command in response and 'SUCCEEDED' in response:
+        return
+      raise SnifferError(full_command, line.strip())
+
+  def _export_tab(self, tab):
+    ending = '_' + tab.lower().rsplit(':', 1)[1].replace(' ', '_') + '.csv'
+    try:
+      self._sniffer_command('Export', 'File=' + self.export_prefix + ending,
+                            'Tab=' + tab)
+    except SnifferError as e:
+      _LOG.warning('Failed to export tab %s: %s', tab, e.response)
+    else:
+      self._files.append(self.export_prefix + ending)
+
+  @property
+  def files(self):
+    """Iterator over the names of log files created by the sniffer."""
+    for path in self._files:
+      yield os.path.basename(path)
+
+  @property
+  def export_prefix(self):
+    """Generate a filename prefix specific to a single sniffer run."""
+    return os.path.join(self._log_path, 'sniff_{}_{}'.format(
+        self.capture_start_time.strftime('%Y%m%d-%H%M%S'), self.capture_id))
+
+  def start_sniffing(self,
+                     bluetooth_mac,
+                     capture_id,
+                     binary_path=REMOTE_BIN_PATH):
+    """Configure the sniffer and start sniffing."""
```

Comment:
args section?  this is a pretty important public api so probably worth being a bit more thorough on this one docstring

Reply:
Done.